Godeps/_workspace/src/github.com/jbenet/go-datastore/query/query.go

```diff
@@ -0,0 +1,254 @@
+operations performed when querying. To avoid pasting here years of
+database research, let’s summarize the operations datastore supports.
+
+Query Operations:
+
+  * namespace - scope the query, usually by object type
+  * filters - select a subset of values by applying constraints
+  * orders - sort the results by applying sort conditions
+  * limit - impose a numeric limit on the number of results
+  * offset - skip a number of results (for efficient pagination)
+
+datastore combines these operations into a simple Query class that allows
+applications to define their constraints in a simple, generic, way without
+introducing datastore specific calls, languages, etc.
+
+Of course, different datastores provide relational query support across a
+wide spectrum, from full support in traditional databases to none at all in
+most key-value stores. Datastore aims to provide a common, simple interface
+for the sake of application evolution over time and keeping large code bases
+free of tool-specific code. It would be ridiculous to claim to support high-
+performance queries on architectures that obviously do not. Instead, datastore
+provides the interface, ideally translating queries to their native form
+(e.g. into SQL for MySQL).
+
+However, on the wrong datastore, queries can potentially incur the high cost
+of performing the aforemantioned query operations on the data set directly in
+Go. It is the client’s responsibility to select the right tool for the job:
+pick a data storage solution that fits the application’s needs now, and wrap
+it with a datastore implementation. As the needs change, swap out datastore
+implementations to support your new use cases. Some applications, particularly
+in early development stages, can afford to incurr the cost of queries on non-
+relational databases (e.g. using a FSDatastore and not worry about a database
+at all). When it comes time to switch the tool for performance, updating the
+application code can be as simple as swapping the datastore in one place, not
+all over the application code base. This gain in engineering time, both at
+initial development and during later iterations, can significantly offset the
+cost of the layer of abstraction.
+
+*/
+type Query struct {
+	Prefix   string   // namespaces the query to results whose keys have Prefix
+	Filters  []Filter // filter results. apply sequentially
+	Orders   []Order  // order results. apply sequentially
+	Limit    int      // maximum number of results
+	Offset   int      // skip given number of results
+	KeysOnly bool     // return only keys.
+}
+
+var (
+// ErrCancelled = errors.New("datastore.query: cancelled early")
+)
+
+// NotFetched is a special type that signals whether or not the value
+// of an Entry has been fetched or not. This is needed because
+// datastore implementations get to decide whether Query returns values
+// or only keys. nil is not a good signal, as real values may be nil.
+var NotFetched = struct{}{}
```

Comment:
you shouldnt use a `struct{}{}` as a sentinel value, they dont compare well.

see: http://play.golang.org/p/W-g3wuGUHN